toolchain/lex/tokenized_buffer.h

```diff
@@ -260,55 +260,101 @@ class TokenizedBuffer : public Printable<TokenizedBuffer> {
     int indent;
   };
 
-  struct TokenInfo {
-    TokenKind kind;
+  class TokenInfo {
+   public:
+    auto kind() const -> TokenKind { return TokenKind::FromInt(kind_); }
+    auto has_leading_space() const -> bool { return has_leading_space_; }
+
```

Comment:
This type is somewhat harder to navigate in its new form. Something like this seems like it might help:

```suggestion
  // Stores lexical information about a token: its kind, whether it has a leading
  // space, its byte offset in the file, and a kind-dependent payload.
  class TokenInfo {
   public:
    auto kind() const -> TokenKind { return TokenKind::FromInt(kind_); }
    auto has_leading_space() const -> bool { return has_leading_space_; }
    auto byte_offset() const -> int32_t { return byte_offset_; }
    
    // Accessors/mutators for the payload. At most one payload accessor is valid
    // for a given kind, along with its corresponding mutator (if any).
```

Reply:
Will add comments (but somewhat different ones), PTAL

Reply:
The class comment is great, but I think it would still be helpful to move `byte_offset` up, and add a comment that collectively documents the payload accessors and mutators (especially their preconditions). That keeps `byte_offset` from getting lost at the bottom, and avoids the need to add a comment to each payload method (which the style guide arguably calls for, but I don't think would be helpful here).

Reply:
My problem with reordering the accessors is that as-is they follow the order of storage in tthe struct. And I think that order is important for something whose entire business is careful layout in memory.

And moreover, as I indicated in the class comment, `byte_offset` is importantly different from all the others, and might be usefully separated out of this struct and into some parallel structure. That also motivated me to keep the others grouped together.

Is there another change that would help the byte offset not be lost for you, besides mingling the order? (I'm having trouble coming up with ideas myself I'm afraid, the byte offset doesn't get lost for me.)

Reply:
I can think of a few other options:
- Giving the payload names a common prefix, e.g. `payload_ident_id`, `payload_string_literal_id`. That doesn't help with `byte_offset` getting visually lost, but it at least makes clear which operations are payload operations (and hence makes clear the scope of the collective payload comment). This may not be a net win for the callsites, though.
- Add a method comment on `byte_offset`, as well as the collective payload comment (and then probably comment the other two unconditional accessors, for consistency). That way the comment on `byte_offset` visually flags the end of the payload methods, and calls attention to the fact that there's another method after it. This is arguably what the style guide calls for, but it's not consistent with the toolchain's convention of putting comments on the private members and leaving the accessors uncommented.
- Move the method bodies out of line, and use whitespace to convey the grouping:
   ```cpp
      auto kind() const -> TokenKind;
      auto has_leading_space() const -> bool;

      // Accessors/mutators for the payload. At most one payload accessor is valid
      // for a given kind, along with its corresponding mutator (if any).
      auto ident_id() const -> IdentifierId;
      auto set_ident_id(IdentifierId ident_id) -> void;
      auto string_literal_id() const -> StringLiteralValueId;
      ...
      auto error_length() const;

      auto byte_offset() const -> int32_t;
   ```
   But that's a mixed bag because right now the method bodies act as implicit documentation of which kinds they can be called with, so arguably this should be paired with more explicit documentation of the kind/payload mapping, maybe at the class level.

These all have drawbacks, as does the status quo, but it's probably not worth sinking more time into this question. I'm fine with however you want to deal with it.

Reply:
I like the idea of just adding comments to the accessors.

I've moved and significantly expanded the comments from the fields, and switched the limited field comments to talk about the *storage* peculiarities only, and not as much about the semantic meaning when accessed. I think its a net improvement, bet let me know.